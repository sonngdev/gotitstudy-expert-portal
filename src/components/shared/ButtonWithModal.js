/* eslint-disable react/jsx-props-no-spreading, react/forbid-prop-types */

import React, { useState } from 'react';
import { Button, Modal } from '@gotitinc/design-system';
import PropTypes from 'prop-types';

export function ButtonWithModal({
  buttonProps,
  buttonAction,
  modalSize,
  modalBody,
  onConfirm,
}) {
  const [modalVisible, setModalVisible] = useState(false);
  const showModal = () => setModalVisible(true);
  const hideModal = () => setModalVisible(false);
  const confirmAndClose = () => {
    onConfirm();
    hideModal();
  };

  return (
    <>
      <Button {...buttonProps} onClick={showModal}>
        {buttonAction}
      </Button>

      <Modal size={modalSize} show={modalVisible} onHide={hideModal}>
        <Modal.Header closeButton>
          <Modal.Title>{buttonAction}</Modal.Title>
        </Modal.Header>

        <Modal.Body>
          {modalBody}
        </Modal.Body>

        <Modal.Footer className="u-flex">
          <Button variant="secondary" className="cancel" onClick={hideModal}>
            Cancel
          </Button>
          <Button variant="primary" className="confirm" onClick={confirmAndClose}>
            {buttonAction}
          </Button>
        </Modal.Footer>
      </Modal>
    </>
  );
}

ButtonWithModal.propTypes = {
  buttonProps: PropTypes.object,
  buttonAction: PropTypes.string,
  modalSize: PropTypes.string,
  modalBody: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
  ]),
  onConfirm: PropTypes.func,
};

ButtonWithModal.defaultProps = {
  buttonProps: {},
  buttonAction: 'Action',
  modalSize: 'small',
  modalBody: 'Are you sure?',
  onConfirm: () => {},
};
