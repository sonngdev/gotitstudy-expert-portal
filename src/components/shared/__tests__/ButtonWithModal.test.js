import React from 'react';
import { shallow } from 'enzyme';
import { Modal } from '@gotitinc/design-system';
import { ButtonWithModal } from '../ButtonWithModal';

describe('<ButtonWithModal />', () => {
  let wrapper;
  let onConfirm;

  beforeEach(() => {
    onConfirm = jest.fn();

    wrapper = shallow(
      <ButtonWithModal
        buttonProps={{ className: 'button' }}
        buttonAction="Action"
        modalSize="medium"
        modalTitle="Modal title"
        modalBody={<p>Modal body</p>}
        confirmButtonText="Confirm button"
        onConfirm={onConfirm}
      />,
    );
  });

  afterEach(() => {
    onConfirm.mockClear();
  });

  it('renders', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('contains a button', () => {
    const button = wrapper.find('.button');
    expect(button.exists()).toBe(true);
    expect(button.text()).toBe('Action');
  });

  it('has a modal that is initially hidden', () => {
    const modal = wrapper.find('Modal');
    expect(modal.exists()).toBe(true);
    expect(modal.prop('show')).toBe(false);
  });

  describe('<Modal />', () => {
    beforeEach(() => {
      wrapper.find('.button').simulate('click');
    });

    it('shows up when button is clicked', () => {
      expect(wrapper.find('Modal').prop('show')).toBe(true);
    });

    it('has the given size', () => {
      expect((wrapper.find('Modal').prop('size'))).toBe('medium');
    });

    it('contains a title', () => {
      const title = wrapper.find(Modal.Title);
      expect(title.exists()).toBe(true);
      expect(title.text()).toBe('Modal title');
    });

    it('contains a body', () => {
      const body = wrapper.find(Modal.Body);
      expect(body.exists()).toBe(true);

      const child = body.childAt(0);
      expect(child.name()).toBe('p');
      expect(child.text()).toBe('Modal body');
    });

    it('contains a confirm button', () => {
      const confirm = wrapper.find('Modal').find('.confirm');
      expect(confirm.exists()).toBe(true);
      expect(confirm.text()).toBe('Confirm button');
    });

    it('hides when cancel button is clicked', () => {
      wrapper.find('Modal').find('.cancel').simulate('click');
      expect(wrapper.find('Modal').prop('show')).toBe(false);
    });

    it('calls onConfirm when confirm button is clicked', () => {
      wrapper.find('Modal').find('.confirm').simulate('click');
      expect(onConfirm).toHaveBeenCalled();
    });
  });
});
