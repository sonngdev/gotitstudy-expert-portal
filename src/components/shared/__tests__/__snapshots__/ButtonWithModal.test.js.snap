// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonWithModal /> renders 1`] = `
"<Fragment>
  <Button className=\\"button\\" onClick={[Function: showModal]} nonUppercase={false} variant=\\"primary\\">
    Foo bar
  </Button>
  <Modal size=\\"medium\\" show={false} onHide={[Function: hideModal]} relative={false} transition={{...}}>
    <Modal.Header closeButton={true}>
      <ModalTitle uText650 uTextCenter>
        Foo bar
      </ModalTitle uText650 uTextCenter>
    </Modal.Header>
    <ModalBody uPaddingHorizontalMedium uBackgroundWhite uPaddingTopSmall uPaddingBottomMedium>
      <p>
        Bar baz
      </p>
    </ModalBody uPaddingHorizontalMedium uBackgroundWhite uPaddingTopSmall uPaddingBottomMedium>
    <ModalFooter uBackgroundLightest uPaddingMedium uFlex uAlignItemsCenter uJustifyContentEnd className=\\"u-flex\\">
      <Button variant=\\"secondary\\" className=\\"cancel\\" onClick={[Function: hideModal]} nonUppercase={false}>
        Cancel
      </Button>
      <Button variant=\\"primary\\" className=\\"confirm\\" onClick={[Function: confirmAndClose]} nonUppercase={false}>
        Foo bar
      </Button>
    </ModalFooter uBackgroundLightest uPaddingMedium uFlex uAlignItemsCenter uJustifyContentEnd>
  </Modal>
</Fragment>"
`;
